### Concept Drift Simulation

In this notebook, we simulate concept drift by altering the relationship between input features and the target variable. This experiment demonstrates how machine learning models can degrade when the underlying data generation process changes.

In [1]:
import joblib
import numpy as np
import pandas as pd

from sklearn.metrics import accuracy_score, f1_score

In [4]:
model = joblib.load("calibrated_rf.pkl")
threshold = joblib.load("best_cost_threshold.pkl")

X_test = joblib.load("X_test.pkl")
y_test = joblib.load("y_test.pkl")

In [5]:
y_proba_original = model.predict_proba(X_test)[:,1]
y_pred_original = (y_proba_original >= threshold).astype(int)

acc_original = accuracy_score(y_test, y_pred_original)
f1_original = f1_score(y_test, y_pred_original)

print("Original Accuracy:", acc_original)
print("Original F1:", f1_original)

Original Accuracy: 0.6984732824427481
Original F1: 0.6108374384236454


In [6]:
X_concept = X_test.copy()

X_concept["Sex"] = 1 - X_concept["Sex"]

In [7]:
y_proba_concept = model.predict_proba(X_concept)[:,1]
y_pred_concept = (y_proba_concept >= threshold).astype(int)

acc_concept = accuracy_score(y_test, y_pred_concept)
f1_concept = f1_score(y_test, y_pred_concept)

print("Concept Drift Accuracy:", acc_concept)
print("Concept Drift F1:", f1_concept)

Concept Drift Accuracy: 0.32061068702290074
Concept Drift F1: 0.32061068702290074


In [8]:
print("Accuracy Drop:", acc_original - acc_concept)
print("F1 Drop:", f1_original - f1_concept)

Accuracy Drop: 0.37786259541984735
F1 Drop: 0.2902267514007446


### In this experiment, concept drift was simulated by reversing the relationship between the "Sex" feature and survival outcomes. Since the model heavily relies on this feature, the change significantly disrupted the learned patterns, leading to a substantial performance drop. Accuracy decreased by approximately 38% and F1 score by about 29%. This result illustrates how sensitive machine learning models can be to changes in the underlying data generation process and highlights the importance of continuous monitoring and periodic model retraining in real-world deployments.